internal/utils.go

```diff
@@ -77,25 +77,46 @@ func (l *LockMap) Get(k string) string {
 // Implementation and related tests were taken/inspired by felixge/countermap
 // https://github.com/felixge/countermap/pull/2
 type XSyncMapCounterMap struct {
-	counts *xsync.MapOf[string, *xsync.Counter]
+	counts  [2]*xsync.MapOf[string, *xsync.Counter]
+	mu, muD sync.Mutex
+	index   int
 }
 
 func NewXSyncMapCounterMap() *XSyncMapCounterMap {
-	return &XSyncMapCounterMap{counts: xsync.NewMapOf[string, *xsync.Counter]()}
+	return &XSyncMapCounterMap{
+		counts: [2]*xsync.MapOf[string, *xsync.Counter]{
+			xsync.NewMapOf[string, *xsync.Counter](),
+			xsync.NewMapOf[string, *xsync.Counter](),
+		},
+	}
 }
 
 func (cm *XSyncMapCounterMap) Inc(key string) {
-	val, ok := cm.counts.Load(key)
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	i := cm.index
+	val, ok := cm.counts[i].Load(key)
 	if !ok {
-		val, _ = cm.counts.LoadOrStore(key, xsync.NewCounter())
+		val, _ = cm.counts[i].LoadAndStore(key, xsync.NewCounter())
 	}
 	val.Inc()
 }
 
+func (cm *XSyncMapCounterMap) swap() int {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	old := cm.index
+	cm.index = cm.index ^ 1
+	return old
+}
+
 func (cm *XSyncMapCounterMap) GetAndReset() map[string]int64 {
+	cm.muD.Lock()
```

Comment:
I think rather than having another lock, we should use https://pkg.go.dev/golang.org/x/sync/singleflight#Group.Do It could be more costly but semantically and practically it makes more sense to have only one `GetAndReset` running.

Reply:
This will also guarantee, we don't `swap` it randomly.

Reply:
The thing about `group.Do` is that it will return the same value for every goroutine that calls the same function with the same key. IIUC, this would lead to double/triple/etc counting if we have two/three/etc goroutines calling `GetAndReset()` at the same time. Do you agree?

I've considered using the `shared` return value to decide whether to count the results or not, but that might lead to _undercounting_, as I think `shared` is `true` for ALL goroutines running into each other...